src/main/java/de/tum/in/www1/artemis/service/scheduled/ParticipantScoreSchedulerService.java

```diff
@@ -0,0 +1,367 @@
+            UserRepository userRepository, TeamRepository teamRepository) {
+        this.scheduler = scheduler;
+        this.participantScoreRepository = participantScoreRepository;
+        this.studentScoreRepository = studentScoreRepository;
+        this.teamScoreRepository = teamScoreRepository;
+        this.exerciseRepository = exerciseRepository;
+        this.resultRepository = resultRepository;
+        this.userRepository = userRepository;
+        this.teamRepository = teamRepository;
+    }
+
+    public Map<Integer, ScheduledFuture<?>> getScheduledTasks() {
+        return scheduledTasks;
+    }
+
+    /**
+     * Schedule all outdated participant scores when the service is started.
+     */
+    @PostConstruct
+    public void startup() {
+        scheduleTasks();
+    }
+
+    /**
+     * Before shutdown, cancel all running or scheduled tasks.
+     */
+    @PreDestroy
+    public void shutdown() {
+        // Stop all running tasks, we will reschedule them on startup again
+        scheduledTasks.values().forEach(future -> {
+            future.cancel(true);
+        });
+        scheduledTasks.clear();
+    }
+
+    /**
+     * Every minute, query for modified results and schedule a task to update the participant scores.
+     * We schedule all results that were created/updated since the last run of the cron job.
+     * Additionally, we schedule all participant scores that are outdated/invalid.
+     */
+    @Scheduled(cron = "0 * * * * *")
+    protected void scheduleTasks() {
+        logger.debug("Schedule tasks to process...");
+        SecurityUtils.setAuthorizationObject();
+
+        // Find all results that were added after the last run (on startup: last time we modified a participant score)
+        var latestRun = lastScheduledRun.orElseGet(() -> participantScoreRepository.getLatestModifiedDate().orElse(Instant.now()));
+        // Update last run time before we continue with time-consuming operations
+        lastScheduledRun = Optional.of(Instant.now());
+
+        var resultsToProcess = resultRepository.findAllByLastModifiedDateAfter(latestRun);
+        resultsToProcess.forEach(result -> {
+            if (result.getParticipation() instanceof StudentParticipation studentParticipation) {
+                var lastModified = result.getLastModifiedDate() == null ? Instant.now() : result.getLastModifiedDate();
+                scheduleTask(studentParticipation.getExercise().getId(), studentParticipation.getParticipant().getId(), lastModified);
+            }
+        });
+
+        var participantScoresToProcess = participantScoreRepository.findAllOutdated();
+        participantScoresToProcess.forEach(participantScore -> {
+            if (participantScore instanceof TeamScore teamScore) {
+                scheduleTask(teamScore.getExercise().getId(), teamScore.getTeam().getId(), Instant.now());
```

Comment:
❓If I understand the code correctly you are passing `Instant.now()` here because you want the below `already-up-to-date` check to fail. Just to confirm if I understood this part correctly

Reply:
Yes, exactly.